lib/cdo/contact_rollups/v2/pardot.rb

```diff
@@ -56,59 +59,117 @@ def self.retrieve_new_ids(last_id)
   #
   # You can think of using this method as writing to an output stream, which at the end you
   # have to flush out the remaining content in the stream.
   #
-  # @param [Hash] contact
+  # @param [String] email
+  # @param [Hash] data
   # @param [Boolean] eager_submit
-  # @return [Array] two arrays, one for all submitted prospects and one for Pardot errors
-  def batch_create_prospects(contact, eager_submit = false)
-    submissions = []
-    errors = []
-
-    prospect = self.class.convert_to_prospect_fields contact
+  # @return [Array<Array>] @see process_batch method
+  def batch_create_prospects(email, data, eager_submit = false)
+    prospect = self.class.convert_to_pardot_prospect data.merge(email: email)
     @new_prospects << prospect
 
-    url = self.class.build_batch_url BATCH_CREATE_URL, @new_prospects
-    if url.length > URL_LENGTH_THRESHOLD || @new_prospects.size == MAX_PROSPECT_BATCH_SIZE || eager_submit
-      # TODO: rescue Net::ReadTimeout from submit_prospect_batch and tolerate a certain number of failures.
-      # Don't retry an insert because it will create duplicate Pardot prospects.
-      errors = self.class.submit_batch_request BATCH_CREATE_URL, @new_prospects
-      submissions = @new_prospects
-      @new_prospects = []
-    end
-
-    [submissions, errors]
+    process_batch BATCH_CREATE_URL, @new_prospects, eager_submit
   end
 
-  # Batch-creates prospect in Pardot. Request is sent immediately.
+  # Immediately batch-create the remaining prospects in Pardot.
+  # @return [Array<Array>] @see process_batch method
   def batch_create_remaining_prospects
-    return [], [] unless @new_prospects.present?
+    process_batch BATCH_CREATE_URL, @new_prospects, true
+  end
 
-    errors = self.class.submit_batch_request BATCH_CREATE_URL, @new_prospects
-    submissions = @new_prospects
-    @new_prospects = []
+  # Compiles a batch of prospects and batch-update them in Pardot when batch size
+  # is big enough. If +eager_submit+ is true, updates the batch in Pardot immediately.
+  #
+  # *Warning:* If +eager_submit+ is false (by default), always uses this method together with
+  # its sibling, +batch_update_remaining_prospects+ to flush out the remaining data in the batch.
+  #
+  # @param [String] email
+  # @param [Integer] pardot_id
+  # @param [Hash] old_prospect_data a hash with Pardot prospect keys
+  # @param [Hash] new_contact_data a hash with original contact keys
+  # @param [Boolean] eager_submit
+  # @return [Array<Array>] @see process_batch method
+  def batch_update_prospects(email, pardot_id, old_prospect_data, new_contact_data, eager_submit = false)
+    new_prospect_data = self.class.convert_to_pardot_prospect new_contact_data
+    delta = self.class.calculate_data_delta old_prospect_data, new_prospect_data
+    return [], [] unless delta.present?
+
+    prospect_key = self.class.convert_to_pardot_prospect(email: email, pardot_id: pardot_id)
```

Comment:
I don't understand this, I thought this method returned a hash not a key?

Reply:
yeah it returns a hash containing email and pardot_id key-value, which serve as a composite key to identify a prospect. How about `prospect_mapping` or `prospect_email_pardot_id` or `email_pardot_id_key`? I'm open to naming suggestions :D

Reply:
Maybe `prospect_mapping` and a comment? Non-blocking

Reply:
I settle on a slightly more verbose variable name, `email_pardot_id`. Hope it can reduce confusion.